src/ansys/templates/python/osl_solution/{{cookiecutter.__project_name_slug}}/src/ansys/solutions/{{cookiecutter.__solution_name_slug}}/solution/monitoring_step.py

```diff
@@ -0,0 +1,283 @@
+    def upload_project_data(self, problem_setup_step: ProblemSetupStep) -> None:
+        """Monitor the progress of the optiSLang project and continuously upload project data."""
+        # Creat monitoring directory
+        Path(self.optislang_log_file.path).parent.mkdir(parents=True, exist_ok=True)
+
+        # Connect to optiSLang instance.
+        osl = Optislang(
+            host=problem_setup_step.tcp_server_host,
+            port=problem_setup_step.tcp_server_port,
+            loglevel=problem_setup_step.optislang_log_level,
+            shutdown_on_finished=True
+        )
+
+        # Configure logging.
+        osl_logger = logging.OslLogger(
+            loglevel=problem_setup_step.optislang_log_level,
+            log_to_file=True,
+            logfile_name=self.optislang_log_file.path,
+            log_to_stdout=True,
+        )
+        osl.__logger = osl_logger.add_instance_logger(osl.name, osl, problem_setup_step.optislang_log_level)
+
+        # Monitor project state and upload data.
+        while True:
+            # Get project state
+            self.project_state = osl.project.get_status()
+            osl.log.info(f"Analysis status: {self.project_state}")
+            # Get project status info
+            with open(self.project_status_info_file.path, "w") as json_file: json.dump(osl.get_osl_server().get_full_project_status_info(), json_file)
+            # Read pyoptislang logs
+            self.optislang_logs = read_log_file(self.optislang_log_file.path)
+            # Get actor status info
+            actors_info, actors_status_info = {}, {}
+            for node_info in problem_setup_step.project_tree:
+                if node_info["uid"] == osl.project.root_system.uid:
+                    node = osl.project.root_system
+                else:
+                    node = osl.project.root_system.find_node_by_uid(node_info["uid"], search_depth=-1)
+                actors_info[node.uid] = osl.get_osl_server().get_actor_info(node.uid)
+                if node.get_states_ids():
+                    actors_status_info[node.uid] = []
+                    for hid in node.get_states_ids():
+                        actors_status_info[node.uid].append(
+                            osl.get_osl_server().get_actor_status_info(node.uid, hid)
+                        )
+            with open(self.actors_info_file.path, "w") as json_file: json.dump(actors_info, json_file)
+            with open(self.actors_status_info_file.path, "w") as json_file: json.dump(actors_status_info, json_file)
+            # Upload fields
+            self.transaction.upload(["project_state"])
+            self.transaction.upload(["project_status_info_file"])
+            self.transaction.upload(["actors_info_file"])
+            self.transaction.upload(["actors_status_info_file"])
+            self.transaction.upload(["optislang_logs"])
+            if self.project_state == "FINISHED":
+                break
+            time.sleep(3)
```

Comment:
can we add a comment here to explain why this sleep is needed?

Reply:
done